Change default `X-XSS-Protection` header to '0'

csutter opened the pull request:
### Summary

This header has been deprecated and the XSS auditor it triggered has been removed from all major modern browsers (in favour of Content Security Policy) that implemented this header to begin with (Firefox never did).

[OWASP](https://owasp.org/www-project-secure-headers/#x-xss-protection) suggest setting this header to '0' to disable the default behaviour on old browsers as it can introduce additional security issues.

Added the new behaviour as a framework default from Rails 7.0.

Some other resources discussing this header and why it should be set to '0':
* https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection
* https://scotthelme.co.uk/security-headers-updates/#removing-the-x-xss-protection-header
* https://stackoverflow.com/questions/9090577/what-is-the-http-header-x-xss-protection#answer-57802070

rafaelfranca commented:
👍🏽 this make sense for all new apps but existing apps should decide if they want to change the value or not. So I think we should introduce the new value behind a configuration flag that is default in Rails 7.0 but for any other Rails version we keep the behavior and ask people to change the config to match Rials 7.0 in the `new_frameworks_default` file. Can you change your PR to reflect that plan?

csutter commented:
@rafaelfranca Thanks for the feedback! I've updated the change to be a defaults change instead, starting at `7.0`.

I've considered having a specific configuration flag for this change, but that felt a bit strange – we'd have to start merging things into the default `default_headers` hash, and if users set their own hash we wouldn't want to override it. So I've made it set a new default for the whole hash.

This way we could maybe also add some additional headers/change existing headers for Rails 7.0 to reflect current best practice, and do it all in one go?

zzak commented:
@csutter Could you rebase please?

Also, copying the important part from that page since the link doesn't go directly to it for those curious:

> X-XSS-Protection: Deprecated.
>
> ⚠️ Warning: The X-XSS-Protection header has been deprecated by modern browsers and its use can introduce additional security issues on the client side. As such, it is recommended to set the header as X-XSS-Protection: 0 in order to disable the XSS Auditor, and not allow it to take the default behavior of the browser handling the response. Please use Content-Security-Policy instead.

csutter committed:
Change default X-XSS-Protection header to '0'

This header has been deprecated and the XSS auditor it triggered
has been removed from all major modern browsers (in favour of
Content Security Policy) that implemented this header to begin with
(Firefox never did).

[OWASP](https://owasp.org/www-project-secure-headers/#x-xss-protection)
suggests setting this header to '0' to disable the default behaviour
on old browsers as it can introduce additional security issues.

Added the new behaviour as a framework default from Rails 7.0.

csutter commented:
@rafaelfranca @zzak I've given this another rebase - would love to see it make it into 7.0!

chiragsinghal commented on guides/source/configuring.md:
```diff
@@ -1681,6 +1681,16 @@ Accepts a string for the HTML tag used to wrap attachments. Defaults to `"action
+      "X-Download-Options" => "noopen",
+      "X-Permitted-Cross-Domain-Policies" => "none",
+      "Referrer-Policy" => "strict-origin-when-cross-origin"
+    }
```
nitpick: do we need the extra empty line above this block?

Also, the hash isn't showing up as a code block on the rendered markdown file.
